tests/check_public_api.py

```diff
@@ -0,0 +1,131 @@
+    nm_command = ["nm", "--demangle", "--dynamic", "--defined-only", so_path]
+    nm_output_bytes = subprocess.check_output(nm_command)
+    nm_output = nm_output_bytes.decode(encoding="utf-8")
+    symbols = {line.split()[2] for line in nm_output.splitlines()}
+    return symbols
+
+
+def get_public_headers(prefix):
+    """
+    It is useful to check if all the public headers are actually installed under $PREFIX/include/openscap.
+    :param prefix: CMAKE_INSTALL_PREFIX
+    :return: a set of public headers filepaths
+    """
+    public_headers = set()
+    include_path = os.path.join(prefix, "include", "openscap")
+    for root, dirs, files in os.walk(include_path):
+        for f in files:
+            full_path = os.path.join(root, f)
+            public_headers.add(full_path)
+    return public_headers
+
+
+def get_public_symbols_from_header(header_path):
+    """
+    Parses a header file to find symbols marked by OSCAP_API macro.
+    :param header_path: path to the header file
+    :return: a list of public symbols in the header file
+    """
+    symbols = set()
+    prototype_regex = re.compile(r"^[ \t]*OSCAP_API[^;]*;", re.MULTILINE)
+    with open(header_path, "r") as header_file:
+        content = header_file.read()
+    prototypes = prototype_regex.findall(content)
+    function_name_regex = re.compile(r"([a-zA-Z0-9_]+)\s*\(.*\)")
+    for p in prototypes:
+        p = p.replace("\n", " ")
+        f_match = function_name_regex.search(p)
+        f_name = f_match.group(1) if f_match else "unknown"
+        symbols.add(f_name)
+    return symbols
+
+
+def get_public_symbols_from_headers(headers):
+    """
+    Find public (exported) symbols in header files by finding symbols
+    marked as OSCAP_API.
+    :param headers: a list of public headers file paths
+    :return: a set of public symbols in public headers
+    """
+    symbols = set()
+    for h in headers:
+        symbols.update(get_public_symbols_from_header(h))
+    return symbols
+
+
+def analyze_project_artifacts(src_dir):
+    """
+    Analyze the project and find all the symbols declared as public in
+    public header files and all the symbols actually exported in built
+    library. This is done on a real build and a real installation.
+    The artifacts are installed in a temporary directory which is
+    automatically removed when this function returns.
+    :param src_dir: Path to project sources
+    :return: tuple (header_symbols, so_symbols)
+    """
+    cwd = os.getcwd()
+    with tempfile.TemporaryDirectory() as prefix:
+        build_dir = os.path.join(src_dir, "build")
+        os.chdir(build_dir)
+        cmake_command = ["cmake",
+                         "-DENABLE_PYTHON3=FALSE",
+                         "-DENABLE_PERL=FALSE",
+                         "-DCMAKE_INSTALL_PREFIX=" + prefix,
+                         "-DENABLE_TESTS=FALSE",
+                         ".."]
+        subprocess.run(cmake_command)
```

Comment:
One last feature request similar to my other comment. Can we get this output not-outputted by default? I'd probably redirect stdout+stderr to the same variable, then only output it if the command failed (and then exit this script too with the same return code).

That way you can pipe the output of this script to a file and easily diff the output between different runs without having to edit the output of this script to remove build output.

Reply:
that would be great